Fix: Wait for clear command to execute before reusing terminal

tcvdh opened the pull request:
clear command was in the middle of executing while the main (debugger) command started executing. Adding a small timeout to make sure the clear command executes before allowing the main command to be sent to the terminal.


this would happen when the `settings.json` has these settings:
```JSON
{
  "debug.terminal.clearBeforeReusing": true,
  "terminal.integrated.focusAfterRun": "terminal",
}
```

Fixes #240953



<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/microsoft/vscode/wiki/How-to-Contribute#pull-requests
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->

tcvdh committed:
Fix: Wait for clear command to execute before reusing terminal

#240953

tcvdh committed:
Merge branch 'main' into fix-clear-command

VACLAVHAVLIS commented:
V pohodě, jsem moc spokojen

connor4312 commented:
We should not just blindly add a delay. If shell integration is available (the common case) we should use `TerminalShellIntegration.executeCommand` and avoid having to delay.

tcvdh commented:
@connor4312 if you check the lines above: https://github.com/microsoft/vscode/blob/a948ee5bec485e4dde01abc6d6e289e3f64dbb14/src/vs/workbench/api/node/extHostDebugService.ts#L155-L158
you can see after the ^C there is also a 200ms delay (which was related to an issue exactly the same as this, except about ^C being in the terminal command instead of clear). so i used the same fix for that on this issue. if something else should be done than the delay, the ^C delay should also be deleted and fixed.

tcvdh committed:
Merge branch 'microsoft:main' into fix-clear-command

tcvdh committed:
Add ShellIntegration to clearBeforeReusing

tcvdh commented:
@connor4312 i have updated the PR code, i added the shellIntegration and kept the delay only for when the shellIntegration isn't available. Can you review my PR again to see if this is what you mean and are satisfied with this implementation?

connor4312 commented on src/vs/workbench/api/node/extHostDebugService.ts:
```diff
@@ -159,14 +159,22 @@ export class ExtHostDebugService extends ExtHostDebugServiceBase {
+					}
+
+					if (terminal.shellIntegration) {
+						terminal.shellIntegration.executeCommand(clearCommand);
```
This returns synchronously, but do we need to wait for the completion signal? My assumption would be that when the following `executeCommand` is sent for the debugger, that's enqueued and run after the current command executes.

Tyriar replied:
I think this might be a race condition issue actually. You can see the implementation here:

https://github.com/microsoft/vscode/blob/0ec1d0b7b5e5d61d99da8bfab62037998da9cf19/src/vs/workbench/api/common/extHostTerminalShellIntegration.ts#L212-L229

Notice we call `this._currentExecution.endExecution(undefined);` to force ending of any executions, so what might happen in practice:

1. Send cls to terminal
2. cls execution is started
3. cls execution is ended
4. `Next` execution starts
5. cls finishes, signal is sent to exthost
6. `Next` execution end

It's a little looser than you might expect since we don't necessarily get a strong signal that a particular execution started/ended. We could probably improve this with repro steps, but in the meantime I think you need the await.

tcvdh replied:
await won't change anything though. 'await' has no effect on the type of this expression.

connor4312 replied:
We would have to separately listen to the `onDidEndTerminalShellExecution` event

tcvdh replied:
For some reason the `onDidEndTerminalShellExecution` event never fires, only the `onDidStartTerminalShellExecution`... no clue why or what exactly is happening here. it also crashes the terminal in the same way as the original issue (opens lldb in background but doesn't attach), which makes sense, because it's not able to attach because it's still listening to the `onDidEndTerminalShellExecution` event, which is never firing, thus never attaching the (now running) debugger to vs code. So for now the code from this PR is the only thing that seems to be working as expected without any issues (without the event listner). If you guys have any input or any way to get the event listner to work, do tell, because i might be missing something.

BTW. while checking this out it seems like there is an issue with the `shellIntegration`, but that seems to be a thing for a separate issue. when you put this in settings.json: `"terminal.integrated.shellIntegration.enabled": false` it doesn't actually stop the shellIntegration from enabling. it will stil enable the `shellIntegration` on new terminals and make `terminal.shellIntegration` never be undefined, even when it should never be enabled

Tyriar replied:
> BTW. while checking this out it seems like there is an issue with the shellIntegration, but that seems to be a thing for a separate issue. when you put this in settings.json: "terminal.integrated.shellIntegration.enabled": false it doesn't actually stop the shellIntegration from enabling. it will stil enable the shellIntegration on new terminals and make terminal.shellIntegration never be undefined, even when it should never be enabled

@tcvdh this could be the case when the shell sends shell integration sequences, you have [manually installed it](https://code.visualstudio.com/docs/terminal/shell-integration#_manual-installation) or if you're reloading the window and the terminals get revived (ie. History restored shows up). That setting only determines whether it's attempted to be injected automatically.

tcvdh replied:
@Tyriar that's the thing. i have not manually installed it into my zsh terminal. and i did close and reopen the terminal. and also on all new zsh terminals it's there. (when you hover over the terminal, you can see if it's integrated or not, which it says it is..). So when testing, the `terminal.shellIntegration` is never undefined on the terminal. 

And if you know what is going on with the event or how to properly get it, please tell. im getting the event from `this._terminalShellIntegrationService.onDidEndTerminalShellExecution` in the code, which is not working like it should as i said

Tyriar replied:
@tcvdh many shells emit shell integration sequences which we take and try to use. Is this plain zsh, oh-my-zsh, powerlevel10k, a custom theme? I can try set up what you have to look into what's going on.

tcvdh replied:
im using oh-my-zsh and powerlevel10k. you mean that those will be used by vs code and thus defining `terminal.shellIntegration`?

Tyriar replied:
powerlevel10k is known to have some problems with shell integration. But I heard recently that you updating fixes some problems https://github.com/microsoft/vscode-copilot-release/issues/4814#issuecomment-2659836959

tcvdh replied:
@Tyriar Thank you! I can't believe it's that simple. I was almost going crazy.

Tyriar replied:
For the event, could you try debugging around these areas?

https://github.com/microsoft/vscode/blob/bb766b362c7a12012d97b82bd1538cecd109f5c7/src/vs/workbench/api/common/extHostTerminalShellIntegration.ts#L235-L241

https://github.com/microsoft/vscode/blob/bb766b362c7a12012d97b82bd1538cecd109f5c7/src/vs/workbench/api/browser/mainThreadTerminalShellIntegration.ts#L93-L101

https://github.com/microsoft/vscode/blob/bb766b362c7a12012d97b82bd1538cecd109f5c7/src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts#L384-L391

If the command finished code above isn't getting hit, you can enable the trace log ("log level" in command palette), open the pty host output channel and try repro again. That will tell you all the data coming into the terminal, searching that for `133` and `633` will find the shell integration sequences.

tcvdh replied:
yeah the powerlevel10k update fixed the events apparently. Implementing now and adding a commit in a few minutes.

Tyriar replied:
Good to hear :+1:

tcvdh replied:
@Tyriar @connor4312 i just pushed another commit, please check it out and let me know if this is to your liking.

tcvdh committed:
Merge branch 'microsoft:main' into fix-clear-command

tcvdh committed:
Add event listner to executeCommand to ensure and await execution

Tyriar commented on src/vs/workbench/api/node/extHostDebugService.ts:
```diff
@@ -159,14 +159,33 @@ export class ExtHostDebugService extends ExtHostDebugServiceBase {
+									resolve();
+								}
+							}));
+							ds.add(disposableTimeout(resolve, 5000)); // 5s timeout to ensure we resolve
```
A shorter timeout (500-1000?) should be fine, especially since there's the problem you were hitting.

tcvdh replied:
sounds right

Tyriar reviewed (dismissed):
lgtm

tcvdh committed:
Changed timeout duration
